src/serde/de_tree.rs

```diff
@@ -0,0 +1,327 @@
+                            start,
+                            end: 0,
+                            right_index: 0,
+                        };
+                        r.push(new_obj);
+                        if calculate_tree_hashes {
+                            tree_hashes.push([0; 32])
+                        }
+                        op_stack.push(ParseOpRef::SaveCursor(index));
+                        op_stack.push(ParseOpRef::ParseObj);
+                        op_stack.push(ParseOpRef::SaveIndex(index));
+                        op_stack.push(ParseOpRef::ParseObj);
+                    } else {
+                        let (start, end, atom_offset, tree_hash) = {
+                            if b <= MAX_SINGLE_BYTE {
+                                (
+                                    start,
+                                    start + 1,
+                                    0,
+                                    tree_hash_for_byte(b, calculate_tree_hashes),
+                                )
+                            } else {
+                                let (atom_offset, atom_size) = decode_size_with_offset(f, b)?;
+                                let end = start + (atom_offset as u64) + atom_size;
+                                let h = skip_or_sha_bytes(f, atom_size, calculate_tree_hashes)?;
+                                (start, end, atom_offset as u32, h)
+                            }
+                        };
+                        if calculate_tree_hashes {
+                            tree_hashes.push(tree_hash.expect("failed unwrap"))
+                        }
+                        let new_obj = ParsedTriple::Atom {
+                            start,
+                            end,
+                            atom_offset,
+                        };
+                        cursor = end;
+                        r.push(new_obj);
+                    }
+                }
+                ParseOpRef::SaveCursor(index) => {
+                    if let ParsedTriple::Pair {
+                        start,
+                        end: _,
+                        right_index,
+                    } = r[index]
+                    {
+                        if calculate_tree_hashes {
+                            let h = sha_blobs(&[
+                                &[2],
+                                &tree_hashes[index + 1],
+                                &tree_hashes[right_index as usize],
+                            ]);
+                            tree_hashes[index] = h;
+                        }
+                        r[index] = ParsedTriple::Pair {
+                            start,
+                            end: cursor,
+                            right_index,
+                        };
+                    }
+                }
+                ParseOpRef::SaveIndex(index) => {
+                    if let ParsedTriple::Pair {
```

Comment:
if `r[index]` is *not* a `ParsedTriple::Pair`, that's an internal error, right? it seems like you should assert that it is, rather than silently ignoring it. Same thing goes for the case above

Reply:
Redid these sections with your ideas.